subprojects/model-core/src/main/java/org/gradle/api/internal/provider/MappingProvider.java

```diff
@@ -69,16 +73,8 @@ public ExecutionTimeValue<? extends OUT> calculateExecutionTimeValue() {
             return ExecutionTimeValue.missing();
         } else {
             // not calling ExecutionTimeValue.fixedValue() directly, because the value may have a side effect
-            return ExecutionTimeValue.value(checkTransformedValue(value.toValue().transform(transformer)));
-        }
-    }
-
-    private Value<OUT> checkTransformedValue(Value<OUT> transformedValue) {
-        // check MappingProvider contract with regard to the transform
-        if (transformedValue.isMissing()) {
-            throw new IllegalStateException("The transformer in MappingProvider must always return a value");
+            return ExecutionTimeValue.value(value.toValue().transform(transformer));
```

Comment:
This does not seem to be an equivalent change.
Previously (before the work on this new side effectful provider), here was a call to the `ExecutionTimeValue.fixedValue`. This call implied that the value that is passed in is never null, although I am sure `assert` never runs in production. My change with an extra `checkTransformedValue` method is aimed at making the contract of the `MappingProvider` more pronounced in the code. The contract is that the transformed value must never be a `MISSING` value.

At the same time, the new call to `ExecutionTimeValue.value` simply swallows a missing value, which may lead to some delayed discoveries of bugs.

Reply:
I vaguely remember that this currently works: `provider.map { null }.present == false`. would that be affected by your validation?